src/Cone.cc

```diff
@@ -0,0 +1,188 @@
+};
+
+/////////////////////////////////////////////////
+Cone::Cone()
+  : dataPtr(gz::utils::MakeImpl<Implementation>())
+{
+}
+
+/////////////////////////////////////////////////
+Errors Cone::Load(ElementPtr _sdf)
+{
+  Errors errors;
+
+  this->dataPtr->sdf = _sdf;
+
+  // Check that sdf is a valid pointer
+  if (!_sdf)
+  {
+    errors.push_back({ErrorCode::ELEMENT_MISSING,
+        "Attempting to load a cone, but the provided SDF "
+        "element is null."});
+    return errors;
+  }
+
+  // We need a cone child element
+  if (_sdf->GetName() != "cone")
+  {
+    errors.push_back({ErrorCode::ELEMENT_INCORRECT_TYPE,
+        "Attempting to load a cone geometry, but the provided SDF "
+        "element is not a <cone>."});
+    return errors;
+  }
+
+  {
+    std::pair<double, bool> pair = _sdf->Get<double>(errors, "radius",
+        this->dataPtr->cone.Radius());
+
+    if (!pair.second)
+    {
+      std::stringstream ss;
+      ss << "Invalid <radius> data for a <cone> geometry. "
+         << "Using a radius of "
+         << this->dataPtr->cone.Radius() << ".";
+      errors.push_back({ErrorCode::ELEMENT_INVALID, ss.str()});
+    }
+    this->dataPtr->cone.SetRadius(pair.first);
+  }
+
+  {
+    std::pair<double, bool> pair = _sdf->Get<double>(errors, "length",
+        this->dataPtr->cone.Length());
+
+    if (!pair.second)
+    {
+      std::stringstream ss;
+      ss << "Invalid <length> data for a <cone> geometry. "
+         << "Using a length of "
+         << this->dataPtr->cone.Length() << ".";
+      errors.push_back({ErrorCode::ELEMENT_INVALID, ss.str()});
+    }
+    this->dataPtr->cone.SetLength(pair.first);
+  }
+
+  return errors;
+}
+
+//////////////////////////////////////////////////
+double Cone::Radius() const
+{
+  return this->dataPtr->cone.Radius();
+}
+
+//////////////////////////////////////////////////
+void Cone::SetRadius(double _radius)
+{
+  this->dataPtr->cone.SetRadius(_radius);
+}
+
+//////////////////////////////////////////////////
+double Cone::Length() const
+{
+  return this->dataPtr->cone.Length();
+}
+
+//////////////////////////////////////////////////
+void Cone::SetLength(double _length)
+{
+  this->dataPtr->cone.SetLength(_length);
+}
+
+/////////////////////////////////////////////////
+sdf::ElementPtr Cone::Element() const
+{
+  return this->dataPtr->sdf;
+}
+
+/////////////////////////////////////////////////
+const gz::math::Coned &Cone::Shape() const
+{
+  return this->dataPtr->cone;
+}
+
+/////////////////////////////////////////////////
+gz::math::Coned &Cone::Shape()
+{
+  return this->dataPtr->cone;
+}
+
+std::optional<gz::math::Inertiald> Cone::CalculateInertial(double _density)
+{
+  gz::math::Material material = gz::math::Material(_density);
+  this->dataPtr->cone.SetMat(material);
+
+  auto coneMassMatrix = this->dataPtr->cone.MassMatrix();
+
+  if (!coneMassMatrix)
+  {
+    return std::nullopt;
+  }
+  else
+  {
+    gz::math::Inertiald coneInertial;
```

Comment:
if we choose to set the origin of a cone shape at the center of its bounding cylinder, then the inertial pose should be set in this function

Reply:
alternatively, we could add an API to `gz::math::Cone` that returns a `gz::math::Inertial` with the `-L/4` Z pose value and call that instead of encoding the cone center of mass location here

Reply:
for example, see https://github.com/gazebosim/sdformat/commit/caab32eeec9171de0e77b456b10b19f4ed31f600

Reply:
@scpeters or would we want to just calculate Ix and Iy at axis going through L/2 since that's where our reference for pose is anyhow? Might need my math checked on it but I'm pretty sure it came out to be the exact same interestingly enough as the moment of inertia of rotating with the axis in center of base.
Should be `Ix=Iy=m*(3*(r^2)+2*(h^2))/20` instead of current one through centroid at Ix=Iy=(3/80)*m*(4*(r^2)+(h^2))

Reply:
No, it's important that moment of inertia is expressed with respect to the center of mass for each geometry. If there are multiple geometries in a link using `<inertial auto="true" />`, the [total inertial will be calculated using the Inertial `+` operator](https://github.com/gazebosim/sdformat/blob/sdf14/src/Link.cc#L661-L672), which assumes moment of inertia is expressed relative to center of mass and [uses the parallel axis theorem as part of its calculation of the new lumped inertial properties](https://github.com/gazebosim/gz-math/blob/gz-math7/include/gz/math/Inertial.hh#L303-L337).

Reply:
Ok, so hypothetical use case, I have a cone I define with uniform density and:
- radius = 2.5m 
- height = 6.125m
What is your desired way that someone writes this in SDF for the inertial to work properly (instead of having the shape originate at height/4 and offsetting in Z with the relative quarter height as the user sees fit)?

Reply:
And also @scpeters would you be ok with opening a PR with the desired implementation that I could cherry-pick if for SDF into this commit or if one is needed for gz-math that I could help review in associating with this PR?

Reply:
> Ok, so hypothetical use case, I have a cone I define with uniform density and:
> 
> * radius = 2.5m
> * height = 6.125m
>   What is your desired way that someone writes this in SDF for the inertial to work properly (instead of having the shape originate at height/4 and offsetting in Z with the relative quarter height as the user sees fit)?

With the shape origin at the midpoint of the central axis, this shape's inertial properties could be specified in two ways:

### explicit inertial properties

~~~
<inertial>
  <!-- Z = -L/4 = -6.125/4 = -1.53125 -->
  <pose>0 0 -1.53125 0 0 0</pose>
  <mass>...</mass>
  <inertia>...</inertia>
</inertial>
~~~

### automatic inertial properties

~~~
<inertial auto="true">
  <density>...</density>
</inertial>
~~~

Reply:
> And also @scpeters would you be ok with opening a PR with the desired implementation that I could cherry-pick if for SDF into this commit or if one is needed for gz-math that I could help review in associating with this PR?

Sure I'll work on that shortly

Reply:
I think the only code changes needed are in https://github.com/gazebosim/sdformat/commit/1e91a6549f801ddd8fa2ff76d9c6587a83911ddb. If you don't mind cherry-picking, that would save me some overhead opening a pull request against your fork.

We may want to add some additional documentation clarifying the definition of the origins, but I think that commit has the only changes needed to the code

Reply:
> With the shape origin at the midpoint of the central axis, this shape's inertial properties could be specified in two ways:
> 
> ### explicit inertial properties
> ```
> <inertial>
>   <!-- Z = -L/4 = -6.125/4 = -1.53125 -->
>   <pose>0 0 -1.53125 0 0 0</pose>
>   <mass>...</mass>
>   <inertia>...</inertia>
> </inertial>
> ```
> 
> ### automatic inertial properties
> ```
> <inertial auto="true">
>   <density>...</density>
> </inertial>
> ```

So as a general user if I say had a 3.5kg cone, would I then need to compute the density to get auto-inertial to work?
Is there a method for only providing mass as volume is already known based on provided geometry parameters?

Reply:
> I think the only code changes needed are in [1e91a65](https://github.com/gazebosim/sdformat/commit/1e91a6549f801ddd8fa2ff76d9c6587a83911ddb). If you don't mind cherry-picking, that would save me some overhead opening a pull request against your fork.
> 
> We may want to add some additional documentation clarifying the definition of the origins, but I think that commit has the only changes needed to the code

cherry-picked and pushed up.

Reply:
While I still like option 2 the most as I think it avoids the most amount of changes/manual calculations from a user perspective and all they have to do is visually realize/observe the physical shape offsets are different. I also just want to get cone support in. So since we cherry-picked your desired fix I'll mark this as resolved.

Reply:
> So as a general user if I say had a 3.5kg cone, would I then need to compute the density to get auto-inertial to work? Is there a method for only providing mass as volume is already known based on provided geometry parameters?

There isn't yet a method to do this, but there is a feature request with discussion about how it could be implemented: https://github.com/gazebosim/sdf_tutorials/issues/98. I've had a few other tasks come up so I haven't had a chance to work further on that.

Reply:
> While I still like option 2 the most as I think it avoids the most amount of changes/manual calculations from a user perspective

I'm not sure I understand why option 2 is easier; I'll follow up offline